tiledb/sm/rest/rest_profile.cc

```diff
@@ -0,0 +1,300 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class RestProfile.
+ */
+
+#include <iostream>
+
+#include "rest_profile.h"
+#include "tiledb/common/random/random_label.h"
+#include "tiledb/sm/misc/constants.h"
+
+using namespace tiledb::common;
+using namespace tiledb::common::filesystem;
+
+namespace tiledb::sm {
+
+/* ****************************** */
+/*       PARAMETER DEFAULTS       */
+/* ****************************** */
+
+const std::string RestProfile::DEFAULT_NAME{"default"};
+const std::string RestProfile::DEFAULT_PASSWORD{""};
+const std::string RestProfile::DEFAULT_PAYER_NAMESPACE{""};
+const std::string RestProfile::DEFAULT_TOKEN{""};
+const std::string RestProfile::DEFAULT_SERVER_ADDRESS{"https://api.tiledb.com"};
+const std::string RestProfile::DEFAULT_USERNAME{""};
+
+/* ****************************** */
+/*   CONSTRUCTORS & DESTRUCTORS   */
+/* ****************************** */
+
+RestProfile::RestProfile(const std::string& name, const std::string& homedir)
+    : version_(constants::rest_profile_version)
+    , name_(name)
+    , filepath_(homedir + constants::rest_profile_filepath)
+    , old_filepath_(homedir + constants::cloud_profile_filepath) {
+  // Fstream cannot create directories. If `homedir/.tiledb/` DNE, create it.
+  std::filesystem::create_directories(homedir + ".tiledb");
+
+  // If the local file exists, load the profile with the given name.
+  if (std::filesystem::exists(filepath_)) {
+    load_from_json_file(filepath_);
+  } else {
+    // If the old version of the file exists, load the profile from there
+    if (std::filesystem::exists(old_filepath_)) {
+      load_from_json_file(old_filepath_);
+    }
+  }
+}
+
+RestProfile::RestProfile(const std::string& name) {
+  /**
+   * Ensure the user's $HOME is found.
+   * There's an edge case in which `sudo` does not always preserve the path to
+   * `$HOME`. In this case, the home_directory() API does not throw, but instead
+   * returns `std::nullopt`. As such, we can check for a value in the returned
+   * `std::optional` path of the home_directory and throw an error to the user
+   * accordingly, so they may decide the proper course of action: set the
+   * $HOME path, or perhaps stop using `sudo`.
+   */
+  auto homedir = home_directory();
+  if (homedir.empty()) {
+    throw RestProfileException(
+        "Failed to create RestProfile; $HOME is not set.");
+  }
+
+  RestProfile(name, homedir);
+}
+
+/* ****************************** */
+/*              API               */
+/* ****************************** */
+
+void RestProfile::set_param(
+    const std::string& param, const std::string& value) {
+  // Validate incoming parameter name
+  if (param_values_.count(param) == 0) {
```

Comment:
I think it would be better to replace `count()` with `find()`, similar to what is done in `RestProfile::get_param`. While the performance impact is likely negligible, since we don’t expect the map to be that large, using `find()` makes the intent clearer and avoids some unnecessary lookups.

Reply:
Nice find. I was thinking that `count()` might be better because it will stop looking once it finds something, but it seems`find()` is better overall (according to the below stack overflow).

ref: https://stackoverflow.com/questions/25490357/checking-for-existence-in-stdmap-count-vs-find